cassiopeia/simulator/SpatialLeafSubsampler.py

```diff
@@ -0,0 +1,211 @@
+        interest and returns a tree pruned to contain lineages relevant
+        to only leaves in the sample (the "induced subtree" on the sample).
+        All fields on the original character matrix persist, but maintains
+        character states, meta data, and the dissimilarity map for the sampled
+        cells only.
+
+        Has the option to keep the single edge leading from the root in the
+        induced subtree, if it exists. This edge is often used to represent the
+        time that the root lives before any divisions occur in the phyologeny,
+        and is useful in instances where the branch lengths are critical, like
+        simulating ground truth phylogenies or estimating branch lengths.
+
+        Args:
+            tree: The CassiopeiaTree for which to subsample leaves
+            keep_singular_root_edge: Whether or not to collapse the single edge
+                leading from the root in the subsample, if it exists
+
+        Returns:
+            A new CassiopeiaTree that is the induced subtree on a sample of the
+                leaves in the given tree
+
+        Raises:
+            LeafSubsamplerError if the sample size is <= 0, or larger than the
+                number of leaves in the tree
+        """
+        attribute = self.__attribute_key
+        bounding_box = self.__bounding_box
+        space = self.__space
+        ratio = self.__ratio
+        number_of_leaves = self.__number_of_leaves
+
+        # Check that the tree has spatial information
+        try:
+            coordinates = tree.get_attribute(tree.leaves[0], attribute)
+        except CassiopeiaTreeError:
+            raise LeafSubsamplerError(
+                f"Attribute {attribute} not present in the tree."
+            )
+        
+        if bounding_box is not None:
+            # Check the dimensions of coordinates and bounding box match
+            if len(coordinates) != len(bounding_box):
+                raise LeafSubsamplerError(
+                    f"Dimensions of coordinates ({len(coordinates)}) and "
+                    f"bounding box ({len(bounding_box)}) do not match."
+                )
+            # Subset the leaves to those within the bounding box
+            leaf_keep = []
+            for leaf in tree.leaves:
+                coordinates = tree.get_attribute(leaf, attribute)
+                for i, (min, max) in enumerate(bounding_box):
+                    if coordinates[i] < min or coordinates[i] > max:
+                        break
+                else:
+                    leaf_keep.append(leaf)
+            
+        if space is not None:
+            # Check the dimensions of coordinates and space match
+            if len(coordinates) != len(space.shape):
+                raise LeafSubsamplerError(
+                    f"Dimensions of coordinates ({len(coordinates)}) and "
+                    f"space ({len(space.shape)}) do not match."
+                )
```

Comment:
Do you want to place any checks on the size of the masking array?  

I really like this addition to the sampler and would like to propose an additional behavior based on this performance -- how about we allow spots to be merged together depending on the size of the masking array? In this way, we can have a lot of control over simulating datasets of different resolution by maintaining the same overall shape but increasing / decreasing the size of the array.

Of course there might be weird behavior if for example a particular leaf's coordinates overlaps two or more adjacent spots, but we're not simulating UMIs here so we can just let the leaf appear in these spots it overlaps (as long it overlaps above a certain threshold, perhaps).

Reply:
Yes, I've added a warning for when the max coordinate is much smaller than the shape of the space

I've added spatial downsampling. Let me know what you think. The current implementation does not rescale the coordinated based on the shape of the space parameter so the user will have to rescale the spatial attribute to test many different resolutions. I did this because its more intuitive and I don't want to distort the shape but happy to automatically rescale or add a rescaling flag if you want this functionality.

I don't see how leaf coordinates overlapping adjacent spots is an issue since each leaf is a represented by a point and the spots are non-overlapping. Am I missing something?

Reply:
Great!

I think it makes sense to expect the user to rescale the space coordinates to match the shape of the sampler.

As for overlapping spots/leaves, I think it's an issue right now given the way we are modeling leaves. In the future, we can think about specifying a spherical shape for each leaf to model how a cell would really look in a spatial transcriptomics experiment. When we do, we will have to think about how cells can participate in multiple spots.